minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,635 @@
+            return CartesianCoord(x=x, y=y) if len(k) == 0 else CartesianCoord(x=x, y=y, k=k[0])
+        if len_coord in (4, 5):
+            u, w, *k, j, z = coord
+            for var, val in {"u": u, "j": j}.items():
+                if not val in [0, 1]:
+                    raise ValueError(f"Expected {var} to be in [0, 1], got {val}")
+            return (
+                ZephyrCoord(u=u, w=w, j=j, z=z)
+                if len(k) == 0
+                else ZephyrCoord(u=u, w=w, k=k[0], j=j, z=z)
+            )
+        raise ValueError(f"coord can have length 2, 3, 4 or 5, got {len_coord}")
+
+    @property
+    def zcoord(self) -> ZephyrCoord:
+        """Returns ZephyrCoordinate corresponding to ccoord"""
+        return cartesian_to_zephyr(self._ccoord)
+
+    def is_quo(self) -> bool:
+        """Decides if the ZNode object is quotient"""
+        return (self._ccoord.k is None) and (self._shape.t is None)
+
+    def to_quo(self) -> ZNode:
+        """Returns the quotient ZNode corresponding to self"""
+        qshape = ZShape(m=self._shape.m)
+        qccoord = CartesianCoord(x=self._ccoord, y=self._ccoord)
+        return ZNode(coord=qccoord, shape=qshape, convert_to_z=self.convert_to_z)
+
+    @property
+    def direction(self) -> int:
+        """Returns direction, 0 or 1"""
+        return self._ccoord.x % 2
+
+    def is_vertical(self) -> bool:
+        """Decides whether self is a vertical qubit"""
+        return self.direction == 0
+
+    def is_horizontal(self) -> bool:
+        """Decides whether self is a horizontal qubit"""
+        return self.direction == 1
+
+    def node_kind(self) -> str:
+        """Returns the node kind, vertical or horizontal"""
+        return "vertical" if self.is_vertical() else "horizontal"
+
+    def neighbor_kind(
+        self,
+        other: ZNode,
+    ) -> str | None:
+        """Returns the kind of coupler between self and other,
+        'internal', 'external', 'odd', or None."""
+        if not isinstance(other, ZNode):
+            other = ZNode(other)
+        if self._shape != other._shape:
+            return
+        coord1 = self._ccoord
+        coord2 = other._ccoord
+        x1, y1 = coord1.x, coord1.y
+        x2, y2 = coord2.x, coord2.y
+        if abs(x1 - x2) == abs(y1 - y2) == 1:
+            return "internal"
+        if x1 % 2 != x2 % 2:
+            return
+        if coord1.k != coord2.k:  # odd, external neighbors only on the same k
+            return
+        if self.is_vertical():  # self vertical
+            if x1 != x2:  # odd, external neighbors only on the same vertical lines
+                return
+            diff_y = abs(y1 - y2)
+            return "odd" if diff_y == 2 else "external" if diff_y == 4 else None
+        else:
+            if y1 != y2:  # odd, external neighbors only on the same horizontal lines
+                return
+            diff_x = abs(x1 - x2)
+            return "odd" if diff_x == 2 else "external" if diff_x == 4 else None
```

Comment:
Nit-picky here, but I'd either skip the final else here and add a comment,
```suggestion

        # else, if horizontal
        if y1 != y2:  # odd, external neighbors only on the same horizontal lines
            return
        diff_x = abs(x1 - x2)
        return "odd" if diff_x == 2 else "external" if diff_x == 4 else None
```
or add a final `return None` to clarify that no neighbours found
```suggestion
        else:
            if y1 != y2:  # odd, external neighbors only on the same horizontal lines
                return
            diff_x = abs(x1 - x2)
            return "odd" if diff_x == 2 else "external" if diff_x == 4 else None

    # no neighbors found
    return None
```

Reply:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752